stream_alert_cli/test.py

```diff
@@ -281,24 +284,75 @@ def test_rule(rule_name, test_record, formatted_record):
                 record for the service to be tested
 
         Returns:
-            [bool] boolean indicating if this rule passed
+            [list] alerts that hit for this rule
+            [integer] count of expected alerts for this rule
+            [bool] boolean where False indicates errors occurred during processing
         """
         event = {'Records': [formatted_record]}
 
         expected_alert_count = test_record.get('trigger_count')
         if not expected_alert_count:
-            expected_alert_count = (0, 1)[test_record['trigger']]
+            expected_alert_count = 1 if test_record['trigger'] else 0
+
+        # Run the rule processor. Passing mocked context object with fake
+        # values and False for suppressing sending of alerts
+        processor = StreamAlert(self.context, False)
+        all_valid_logs = processor.run(event)
 
-        # Run the rule processor. Passing 'None' for context
-        # will load a mocked object later
-        alerts = StreamAlert(None, True).run(event)
+        if not all_valid_logs:
+            payload = StreamPayload(raw_record=formatted_record)
+            classifier = StreamClassifier(config=load_config())
+            classifier.map_source(payload)
+            logs = classifier._log_metadata()
+            self.analyze_log_delta(logs, rule_name, test_record)
+
+        alerts = processor.get_alerts()
 
         # we only want alerts for the specific rule being tested
         alerts = [alert for alert in alerts
                   if alert['rule_name'] == rule_name]
 
-        return alerts, expected_alert_count
+        return alerts, expected_alert_count, all_valid_logs
+
+    def analyze_log_delta(self, logs, rule_name, test_record):
+        """Provide some additional context on why this test failed. This will
+        perform some analysis of the test record to determine if keys are
+        missing or if there are unnecessary keys. Any errors are appended to
+        a list of errors so they can be printed at the end of the test run.
 
+        Args:
+            logs [dict]: All of the log schema information for the source/entity
+            rule_name [str]: Name of rule being tested
+            test_record [dict]: Actual record data being tested
+        """
+        rule_info = StreamRules.get_rules()[rule_name]
+        test_record_keys = set(test_record['data'])
+        for log in rule_info.logs:
+            log_keys = set(logs[log]['schema'])
+
+            schema_diff = log_keys.difference(test_record_keys)
+            if schema_diff:
+                pluralize = 's' if len(schema_diff) > 1 else ''
```

Comment:
I'm in favor of removing the multiple pluralize lines of code across our files, for the use of parenthesis and no string interpolation. Ex: `'Data is invalid due to missing key(s) in test record: ...`